File: samples/languages/java/src/main/java/com/lowbudgetlcs/riot4k/samples/java/AccountRepository.java
package com.lowbudgetlcs.riot4k.samples.java;

import com.lowbudgetlcs.riot4k.api.java.Riot4KAsync;
import com.lowbudgetlcs.riot4k.core.result.RiotResult;
import com.lowbudgetlcs.riot4k.core.routes.RegionalRoute;
import com.lowbudgetlcs.riot4k.models.account.v1.AccountDto;

import java.util.concurrent.CompletableFuture;

/**
 * The application's account access layer: maps SDK results onto the app's own
 * {@link AccountLookup} states so the rest of the app never handles raw API types.
 */
public final class AccountRepository {
    private final Riot4KAsync riot4k;

    public AccountRepository(Riot4KAsync riot4k) {
        this.riot4k = riot4k;
    }

    public CompletableFuture<AccountLookup> lookup(RegionalRoute route, String gameName, String tagLine) {
        return riot4k.accountV1()
            .getByRiotIdAsync(route, gameName, tagLine)
            .thenApply(result -> map(result, gameName, tagLine));
    }

    private static AccountLookup map(RiotResult<AccountDto> result, String gameName, String tagLine) {
        if (result instanceof RiotResult.Success<AccountDto> success) {
            AccountDto account = success.getData();
            String riotId = (account.getGameName() != null ? account.getGameName() : gameName)
                + "#" + (account.getTagLine() != null ? account.getTagLine() : tagLine);
            return new AccountLookup.Found(account.getPuuid(), riotId);
        }
        if (result instanceof RiotResult.NotFound) {
            return new AccountLookup.Missing();
        }
        RiotResult.Failure failure = (RiotResult.Failure) result;
        return new AccountLookup.Error(failure.getMessage());
    }
}

File: samples/languages/typescript/src/accountRepository.ts
import { Riot4KJs } from "riot4k-api";

/** Application-level outcome of an account lookup. */
export type AccountLookup =
  | { kind: "found"; puuid: string; riotId: string }
  | { kind: "missing" }
  | { kind: "error"; message: string };

/**
 * The application's account access layer: maps SDK results onto the app's own
 * states so the rest of the app never handles raw API types.
 */
export class AccountRepository {
  constructor(private readonly riot4k: Riot4KJs) {}

  async lookup(route: string, gameName: string, tagLine: string): Promise<AccountLookup> {
    const result = await this.riot4k.accountV1().getByRiotId(route, gameName, tagLine);
    switch (result.type) {
      case "success": {
        const account = result.account;
        if (account == null) {
          return { kind: "error", message: "success result without an account" };
        }
        return {
          kind: "found",
          puuid: account.puuid,
          riotId: `${account.gameName ?? gameName}#${account.tagLine ?? tagLine}`,
        };
      }
      case "notFound":
        return { kind: "missing" };
      default:
        return {
          kind: "error",
          message: result.message ?? `request failed with status ${result.statusCode}`,
        };
    }
  }
}
